src/main/java/org/spongepowered/api/event/message/MessageEvent.java

```diff
@@ -24,37 +24,254 @@
  */
 package org.spongepowered.api.event.message;
 
+import static com.google.common.base.Preconditions.checkNotNull;
+import static org.spongepowered.api.text.TextTemplate.arg;
+import static org.spongepowered.api.text.TextTemplate.of;
+
 import org.spongepowered.api.event.Event;
+import org.spongepowered.api.event.impl.AbstractMessageEvent;
+import org.spongepowered.api.eventgencore.annotation.ImplementedBy;
+import org.spongepowered.api.eventgencore.annotation.PropertySettings;
 import org.spongepowered.api.text.Text;
-
-import java.util.Optional;
-
-import javax.annotation.Nullable;
+import org.spongepowered.api.text.TextRepresentable;
+import org.spongepowered.api.text.transform.FixedPartitionedTextFormatter;
+import org.spongepowered.api.text.transform.SimpleTextFormatter;
+import org.spongepowered.api.text.transform.SimpleTextTemplateApplier;
+import org.spongepowered.api.text.transform.TextTemplateApplier;
 
 /**
  * Describes events when a involving a {@link Text} message.
  */
+@ImplementedBy(AbstractMessageEvent.class)
 public interface MessageEvent extends Event {
 
     /**
+     * Parameter for header section in header partition.
+     */
+    String PARAM_MESSAGE_HEADER = "header";
+
+    /**
+     * Parameter for body section in body partition.
+     */
+    String PARAM_MESSAGE_BODY = "body";
+
+    /**
+     * Parameter for footer section in footer partition.
+     */
+    String PARAM_MESSAGE_FOOTER = "footer";
+
+    /**
      * Gets the original {@link Text} message.
      *
-     * @return The message if present, otherwise {@link Optional#empty()}
+     * @return The message
+     */
+    @PropertySettings(requiredParameter = false)
+    Text getOriginalMessage();
+
+    /**
+     * Returns true if the {@link Text} returned by
+     * {@link #getMessage()} should not be sent.
+     *
+     * @return True if message should not be sent
+     */
+    boolean isMessageCancelled();
```

Comment:
There are events where you might want to cancel just the message and not the entire event itself. i.e. `KickPlayerEvent`